weed/shell/command_volume_tier_move.go

```diff
@@ -82,15 +87,87 @@ func (c *commandVolumeTierMove) Do(args []string, commandEnv *CommandEnv, writer
 	fmt.Printf("tier move volumes: %v\n", volumeIds)
 
 	_, allLocations := collectVolumeReplicaLocations(topologyInfo)
+	allLocations = filterLocationsByDiskType(allLocations, toDiskType)
+	keepDataNodesSorted(allLocations, toDiskType)
+	
+	if len(allLocations) > 0 &&  *limitWorkers > 0 && *limitWorkers < len(allLocations) {
+		allLocations = allLocations[:*limitWorkers]
+	}
+
+	wg := sync.WaitGroup{}
+	bufferLen := len(allLocations)
+	c.queues = make(map[pb.ServerAddress]chan volumeTierMoveJob)
+	
+	for _, dst := range allLocations {
+		destServerAddress := pb.NewServerAddressFromDataNode(dst.dataNode)
+		c.queues[destServerAddress] = make(chan volumeTierMoveJob, bufferLen)
+
+		wg.Add(1)
+		go func (dst location, jobs <-chan volumeTierMoveJob, applyChanges bool) {
+			for job := range jobs {
+				fmt.Fprintf(writer, "moving volume %d from %s to %s with disk type %s ...\n", job.vid, job.src, dst.dataNode.Id, toDiskType.ReadableString())
+				if !applyChanges {
+					continue
+				}
+
+				locations, found := commandEnv.MasterClient.GetLocations(uint32(job.vid))
+				if !found {
+					fmt.Printf("volume %d not found", job.vid)
+					continue
+				}
+
+				unlock := c.Lock(job.src)
+
+				if err := c.doMoveOneVolume(commandEnv, writer, job.vid, toDiskType, locations, job.src, dst); err != nil {
+					fmt.Fprintf(writer, "move volume %d %s => %s: %v\n", job.vid, job.src, dst.dataNode.Id, err)
+				}
+				unlock()
+			}
+			wg.Done()
+		}(dst, c.queues[destServerAddress], *applyChange)
+	}
+
 	for _, vid := range volumeIds {
-		if err = c.doVolumeTierMove(commandEnv, writer, vid, toDiskType, allLocations, *applyChange); err != nil {
+		if err = c.doVolumeTierMove(commandEnv, writer, vid, toDiskType, allLocations); err != nil {
 			fmt.Printf("tier move volume %d: %v\n", vid, err)
 		}
+		allLocations = rotateDataNodes(allLocations)
```

Comment:
This PR is in the right direction. Thanks! It is already excellent since you are just started with Go.

The algo needs some changes. Here the destinations are not "weighted". Some could be "more empty" than others, although they are already the top-k empty destinations.

How about calling "keepDataNodesSorted()" again after planning the move and adjust the estimated volume count?

Reply:
The problem with calling `keepDataNodesSorted()` on each turn is within those "more empty" volumes. Say if we have a server with 1300 volumes and another with 1400, the first one would be choosen on each turn. This practically enqueue all migration job to just one server, turning the algo back to fully sequential.  This is why I'm calling `rotate`, just to ensure that we choose next server on the next turn.

The algo by itself is pretty much dumb, because two neighbor queues might have a job with the same source server effectivily blocking each other and turning it sequentual again. So we just pushing jobs to queues and hoping for the best. Idk if we can improve it somehow, at least for single pass algo. May be we could sort vids somehow to distribute them better to queues. 

Also, I'm unsure about increasing of the volume count in the enqueueing cycle, because they are also updated after succesful move routines. Could led to some edge cases with "almost full" destinations.

Reply:
This algo let all nodes in the "allLocations" to receive about equal number of volumes. So some of the less empty nodes may have more than average number of volumes.

Should be ok to let the more empty nodes to have more volumes to move into. It will take more time to run. But it is better than waiting for the next pass?

Reply:
My point is that if you say just added a new volume server to a destination tier, all jobs would be enqueued to it, making the whole migration process fully sequentual again. This is why I forcing the algo to pick next server on the next turn. Can you suggest how to make it weighted? 
I think we can add tier filter to `volume.balance` to address that problem though.

Reply:
> added a new empty volume server to a destination tier, all jobs would be enqueued to it

This is the ideal way to place the volumes, correct? Waiting for the `volume.balance` is ok, but not so efficient.

Reply:
Yeah, maybe. But it kills the whole point of parallelism, at list for a couple of my setups. One of them has no replicas enabled and for some reason usually have like 10 to 100 volume difference between servers. As far as it is also suffering from slow migrations, real parallelism most likely won't happen anytime soon. Adding a new server and waiting for it fill up to 1500 volumes would also be painfully slow.

Reply:
I see. You want to parallelize it before reaching the balanced state. Make a lot of sense!

Reply:
I am going to merge this,and adjust some minor things, such as variable names, etc.